### PR TITLE
[Fix] `no-invalid-html-attribute`: convert autofix to suggestion

### PR DESCRIPTION
**Pull request by himanshu007-creator**
Fixes : #3458

### CODE REVIEW

File: lib/rules/no-invalid-html-attribute.js [hunk 1]
```diff
@@ -254,9 +254,14 @@ function checkLiteralValueNode(context, attributeName, node, parentNode, parentN
     report(context, messages.onlyStrings, 'onlyStrings', {
       node,
       data: { attributeName },
-      fix(fixer) {
-        return fixer.remove(parentNode);
-      },
+      suggest: [
+        {
+            desc: "To Be Replaced By Appropriate Message",
+            fix: function(fixer) {
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:260]:
```suggestion
            fix(fixer) {
```

File: lib/rules/no-invalid-html-attribute.js [hunk 2]
```diff
@@ -265,9 +270,14 @@ function checkLiteralValueNode(context, attributeName, node, parentNode, parentN
     report(context, messages.noEmpty, 'noEmpty', {
       node,
       data: { attributeName },
-      fix(fixer) {
-        return fixer.remove(parentNode);
-      },
+      suggest: [
+        {
+            desc: "To Be Replaced By Appropriate Message",
+            fix: function(fixer) {
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:276]:
```suggestion
            fix(fixer) {
```

File: lib/rules/no-invalid-html-attribute.js [hunk 3]
```diff
@@ -283,9 +293,14 @@ function checkLiteralValueNode(context, attributeName, node, parentNode, parentN
           attributeName,
           reportingValue,
         },
-        fix(fixer) {
-          return fixer.removeRange(singlePart.range);
-        },
+        suggest: [
+          {
+              desc: "To Be Replaced By Appropriate Message",
+              fix: function(fixer) {
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:299]:
```suggestion
              fix(fixer) {
```

File: lib/rules/no-invalid-html-attribute.js [hunk 4]
```diff
@@ -362,6 +392,17 @@ function checkAttribute(context, node) {
     return fixer.remove(node);
   }
 
+  function suggestion(desc, fixer){
+    return [
+      {
+          desc: desc,
+          fix: function(fixer) {
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:399]:
```suggestion
  function suggestion(desc, fixer) {
    return [
      {
          desc: desc,
          fix(fixer) {
```

File: lib/rules/no-invalid-html-attribute.js [hunk 5]
```diff
@@ -337,17 +357,27 @@ function checkLiteralValueNode(context, attributeName, node, parentNode, parentN
...
-        fix(fixer) {
-          return fixer.replaceTextRange(whitespacePart.range, '\u0020');
-        },
+        suggest: [
+          {
+              desc: "To Be Replaced By Appropriate Message",
+              fix: function(fixer) {
+                  return fixer.removeRange(whitespacePart.range, '\u0020');
+              }
+          }
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:379]: seems like these should all use the `suggestion` function as well?

**Review by ljharb — COMMENTED** [review]
This seems pretty reasonable - with test changes, and the appropriate descriptions, this should be good to go

File: lib/rules/no-invalid-html-attribute.js [hunk 6]
```diff
@@ -234,6 +234,21 @@ const messages = {
...
+}
+
+const fixMessages = {
+  removeDefault : '"remove {{attributeName}}"',
+  removeEmpty:'"remove empty attribute {{attributeName}}"',
+  removeInvalid: '“remove invalid attribute {{reportingValue}}”',
+  removeWhitespaces: 'remove whitespaces in “{{reportingValue}}”',
+
+
+}
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:250]:
```suggestion
};

const fixMessages = {
  removeDefault : '"remove {{attributeName}}"',
  removeEmpty:'"remove empty attribute {{attributeName}}"',
  removeInvalid: '“remove invalid attribute {{reportingValue}}”',
  removeWhitespaces: 'remove whitespaces in “{{reportingValue}}”',
};
```

File: lib/rules/no-invalid-html-attribute.js [hunk 7]
```diff
@@ -358,8 +361,25 @@ const DEFAULT_ATTRIBUTES = ['rel'];
...
+  function fix(fixer, mode, ...args) {
+    if (mode === 'removeTextRange') {
+      return fixer.replaceTextRange(args)
+    }
+    else if (mode === 'removeRange') {
+      return fixer.removeRange(args)
+    }
+    else {
+      return fixer.remove(args)
+    }
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:373]:
```suggestion
    }
    if (mode === 'removeRange') {
      return fixer.removeRange(args);
    }
    return fixer.remove(args);
```

**Review by ljharb — COMMENTED** [review]
Looks good! Let's update the tests and get this landed :-)

**Comment by ljharb** [comment]
@himanshu007-creator any test updates? there's a bunch of failing tests :-)

**Comment by himanshu007-creator** [comment]
Hi @ljharb , i would require some help on this

**Comment by ljharb** [comment]
I think what's needed is that all the test cases that were asserting `output` - ie, an autofixer - instead need to assert suggestions with NO output.

I'm also seeing a syntax error in the logs, which I'll rereview for.

File: lib/rules/no-invalid-html-attribute.js [hunk 8]
```diff
@@ -254,9 +267,7 @@ function checkLiteralValueNode(context, attributeName, node, parentNode, parentN
     report(context, messages.onlyStrings, 'onlyStrings', {
       node,
       data: { attributeName },
-      fix(fixer) {
-        return fixer.remove(parentNode);
-      },
+      suggestion(fixMessages.removeNonString, fixer, fixModes.remove, parentNode)
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:270]: all these lines are invalid syntax - what does `suggestion` return? it seems like maybe the definition of that function is missing, but it'd need to return an object key's *value*, or, an object that'd need to be spread in here.

Comment (himanshu007-creator) [reply lib/rules/no-invalid-html-attribute.js]: yeah, tried something on these lines

File: lib/rules/no-invalid-html-attribute.js [hunk 9]
```diff
@@ -234,6 +234,34 @@ const messages = {
...
+    return fixer.replaceTextRange(args)
+  }
+  if (mode === 'removeRange') {
+    return fixer.removeRange(args)
+  }
+    return fixer.remove(args)
+}
+
+function suggestion(desc, mode, ...args) {
+  return {desc: desc, ...fix(mode, args)}
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:261]:
```suggestion
  return { desc: desc, fix: fix(mode, args) };
```

File: lib/rules/no-invalid-html-attribute.js [hunk 10]
```diff
@@ -234,6 +234,34 @@ const messages = {
...
+}
+
+function fix (fixer, mode, ...args) {
+  if (mode === 'removeTextRange') {
+    return fixer.replaceTextRange(args)
+  }
+  if (mode === 'removeRange') {
+    return fixer.removeRange(args)
+  }
+    return fixer.remove(args)
```
Comment (ljharb) [inline lib/rules/no-invalid-html-attribute.js:257]:
```suggestion
  return () => {
    if (mode === 'removeTextRange') {
      return fixer.replaceTextRange(args);
    }
    if (mode === 'removeRange') {
      return fixer.removeRange(args);
    }
    return fixer.remove(args);
  };
```

**Comment by ljharb** [comment]
ok, i think this will probably get most of them passing

**Comment by ljharb** [comment]
ok, now the failures are expected :-) all the "output" test cases will need to be updated to *not* have output, but to have suggestions.

**Comment by ljharb** [comment]
Looks better! Seems like for the error cases, each error needs to supply a suggestions array.

**Comment by himanshu007-creator** [comment]
@ljharb finally tests passed! :)

**Review by ljharb — APPROVED** [review]

**Pull request merged**